agent/discover/cluster.go

```diff
@@ -0,0 +1,91 @@
+package discover
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+
+	// Reusing the Prometheus Ha Exporter cibadmin xml parser here
+	"github.com/ClusterLabs/ha_cluster_exporter/collector/pacemaker/cib"
+
+	consul "github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+	consul_internal "github.com/trento-project/trento/internal/consul"
+)
+
+const crm_monPath string = "/usr/sbin/crm_mon"
+const cib_admPath string = "/usr/bin/cibadmin"
+
+//const crm_monPath string = "./test/fake/crm_mon.sh"
+//const cib_admPath string = "./test/fake_cibadmin.sh"
+
+// This Discover handles any Pacemaker Cluster type
+type ClusterDiscover struct {
+	cibConfig      cib.Root
+	clusterName    string
+	stonithEnabled bool
+	stonithType    consul_internal.ClusterStonithType
+}
+
+// check if the current node this trento agent is running on can be discovered
+// by ClusterDiscover
+func (cluster ClusterDiscover) shouldDiscover(client consul.Client) bool {
+	// ### Check if we have cibadmin available
+	return true
+}
+
+// Provide a Consul Key-Value Store Prefix that is used to store individual Key-Values for this Cluster
+func (cluster ClusterDiscover) KVStorePrefix() string {
+	return fmt.Sprintf(
+		"%s/%s",
+		consul_internal.KV_CLUSTERS_PATH, cluster.clusterName)
+}
+
+func NewClusterDiscover(client consul.Client) Discovery {
+
+	return func() (Discoverer, error) {
+		r := ClusterDiscover{}
+		r.stonithType = consul_internal.CLUSTER_STONITH_UNKNOWN
+
+		cibParser := cib.NewCibAdminParser(cib_admPath)
+
+		cibConfig, err := cibParser.Parse()
+		if err != nil {
+			log.Printf("Failing to parse: %s", errors.Wrap(err, "crm_mon parser error"))
+		}
+		r.cibConfig = cibConfig
+
+		for _, prop := range r.cibConfig.Configuration.CrmConfig.ClusterProperties {
+			switch prop.Id {
+			case "cib-bootstrap-options-cluster-name":
+				r.clusterName = prop.Value
+			case "cib-bootstrap-options-stonith-enabled":
+				r.stonithEnabled = bool(prop.Value == "true")
+			}
+		}
+
+		var foundVIP bool
+		for _, primitive := range r.cibConfig.Configuration.Resources.Primitives {
+			switch primitive.Type {
+			case "external/sbd":
+				r.stonithType = consul_internal.CLUSTER_STONITH_SBD
+			case "IPaddr2":
+				if !foundVIP && primitive.Provider == "heartbeat" {
+					for _, attr := range primitive.InstanceAttributes {
+						switch attr.Name {
+						case "ip":
+							StoreDiscovery(r, client, "VIP_primary", string(attr.Value))
```

Comment:
I wouldn't store this information this way. I think we should store all the resources entries in a single KV entrypoint as json, so we can just read everything in one singles read.
Anyway, it is fine by now. We can evolve in the future if needed.

Reply:
there is a size limit on kv store values, so storing everything as "json" under a single key is not working either. 

As far as I know you can fetch all key value pairs in one call from the kvstore, and filter then in your view. so that approach should be working. if it is for some reason too slow or too cumbersome we can change it.

Reply:
Well, the limitation is `512kb`. In any case, you are right, you can all elements from a prefix in an individual call. 
Up to you. I though that it would be much easier to marshall and unmarshall things as json, but if you think that the size limitation impacts, we cannot use it.

Reply:
@arbulu89 nah, let's avoid putting json inside a json. I tried that with the checker result output, and it blew up on me royally! 😁